clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,268 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(SaslClientAuthenticator.class);
+    private SaslClient saslClient;
+    private Subject subject;
+    private String servicePrincipal;
+    private PrincipalBuilder principalBuilder;
+    private String host;
+    private String node;
+    private TransportLayer transportLayer;
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+    private byte[] saslToken = new byte[0];
+
+    public enum SaslState {
+        INITIAL, INTERMEDIATE, COMPLETE, FAILED
+    }
+
+    private SaslState saslState = SaslState.INITIAL;
+
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host) throws IOException {
+        this.node = node;
+        this.subject = subject;
+        this.host = host;
+        this.servicePrincipal = servicePrincipal;
+        this.saslClient = createSaslClient();
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder) throws KafkaException {
+        this.transportLayer = transportLayer;
+        this.principalBuilder = principalBuilder;
+        this.saslClient = createSaslClient();
+    }
+
+    private SaslClient createSaslClient() {
+        final Object[] principals = subject.getPrincipals().toArray();
+        // determine client principal from subject.
+        final Principal clientPrincipal = (Principal) principals[0];
+        final KerberosName clientKerberosName = new KerberosName(clientPrincipal.getName());
+        // assume that server and client are in the same realm (by default; unless the config
+        // "kafka.server.realm" is set).
+        String serverRealm = System.getProperty("kafka.server.realm", clientKerberosName.realm());
+        KerberosName serviceKerberosName = new KerberosName(servicePrincipal + "@" + serverRealm);
+        final String clientPrincipalName = clientKerberosName.toString();
+        try {
+            saslClient = Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
+                    public SaslClient run() throws SaslException {
+                        LOG.debug("Client will use GSSAPI as SASL mechanism.");
+                        String[] mechs = {"GSSAPI"};
+                        LOG.debug("creating sasl client: client=" + clientPrincipalName + ";service=" + servicePrincipal + ";serviceHostname=" + host);
+                        SaslClient saslClient = Sasl.createSaslClient(mechs, clientPrincipalName, servicePrincipal, host, null, new ClientCallbackHandler(null));
+                        return saslClient;
+                    }
+                });
+            return saslClient;
+        } catch (Exception e) {
+            LOG.error("Exception while trying to create SASL client", e);
+            throw new KafkaException("Failed to create SASL client", e);
+        }
+    }
+
+
+    public void authenticate() throws IOException {
```

Comment:
Could we document the handshake protocol in the comment? It seems that for each token, we first send a 4-byte size, followed by the bytes in the token itself?